### PR TITLE
Adding new managed WordPress hosts to be identified in class-functions.php.

### PR DESCRIPTION
**Pull request by oikeme**
Adding more hosting providers to get_hosting_provider() function to be able to identify managed WordPress hosts.

Fixes N/A

#### Changes proposed in this Pull Request:

* Adding options to identify new hosts specifically Managed WordPress hosts.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* New Feature pbtFPC-1M-p2

#### Testing instructions:

* Unit Tests: phpunit --filter=WP_Test_Jetpack_Sync_Functions
#### Proposed changelog entry for your changes:

* Add more managed WordPress hosts to hosting provider detection.

### CODE REVIEW

File: packages/sync/src/class-functions.php [hunk 1]
```diff
@@ -161,15 +161,21 @@ public static function get_post_type_features() {
...
+		$hosting_provider_constants = array(
+			'GD_SYSTEM_PLUGIN_DIR' => 'gd-managed-wp',
+			'MM_BASE_DIR'          => 'bh',
+			'PAGELYBIN'            => 'pagely',
+			'KINSTAMU_VERSION'     => 'kinsta',
+			'FLYWHEEL_CONFIG_DIR'  => 'flywheel',
+			'IS_PRESSABLE'         => 'pressable',
+		);
+
+		foreach ( $hosting_provider_constants as $constant => $constant_value ) {
```
Comment (ebinnion) [inline packages/sync/src/class-functions.php]: Instead of `$constant_value`, perhaps we can use `$hosting_provider`?

Comment (oikeme) [reply packages/sync/src/class-functions.php]: Makes more sense. I've updated this

File: tests/php/sync/test_class.jetpack-sync-callables.php [hunk 2]
```diff
@@ -1053,6 +1053,50 @@ public function test_sync_callable_recursive_gets_checksum() {
...
 
+	/**
+	 * Test get_hosting_provider() callable to ensure that known hosts have the
+	 * right hosting provider returned.
+	 *
+	 * @return void
+	 */
+	public function test_get_hosting_provider_callable() {
+		Constants::set_constant( 'GD_SYSTEM_PLUGIN_DIR', 'set' );
+		if ( Constants::is_defined( 'GD_SYSTEM_PLUGIN_DIR' ) || class_exists( '\\WPaaS\\Plugin' ) ) {
```
Comment (tyxla) [inline tests/php/sync/test_class.jetpack-sync-callables.php]: Drive-by comment: it feels a bit odd to have those checks in tests. I'd expect that the `::is_defined` check is not necessary because we just set the constant on the previous line; and instead of the `class_exists` check I'd suggest defining the class or even better - wrapping the `class_exists` check in another method that could be mocked for testing purposes.

**Comment by ebinnion** [comment]
This is looking pretty good to me. I agree with @tyxla's feedback on the the tests.

Besides that, I wouldn't worry much about the Code Climate issue. It seems like it doesn't like the length of the method, but that may be difficult to fix and frankly, the method will continue to grow as we we add hosts in the future. I would suggest updating/removing the functional testing instructions since you've added test coverage.

File: tests/php/sync/test_class.jetpack-sync-callables.php (hunk 2 above)
Comment (oikeme) [reply tests/php/sync/test_class.jetpack-sync-callables.php]: @tyxla thanks for the suggestion on this! It led me to make a couple of other changes to including splitting out the different types of checks and updating the tests.

**Review by ebinnion — DISMISSED** [review]
This LGTM. But, I'd like to get one other set of eyes on it as well.

**Comment by oikeme** [comment]
@tyxla I've made some changes to this patch since your suggestion which was great by the way. Do you mind giving it a look over again? Thanks!

File: packages/sync/src/class-functions.php [hunk 3]
```diff
@@ -161,22 +161,90 @@ public static function get_post_type_features() {
...
+	/**
+	 * Return a hosting provider using a set of known constants.
+	 *
+	 * @return mixed A host identifier string or false.
+	 */
+	public function get_hosting_provider_by_known_constant() {
+		$hosting_provider = false;
+
+		switch ( true ) {
+			case ( Constants::is_defined( 'GD_SYSTEM_PLUGIN_DIR' ) ):
```
Comment (tyxla) [inline packages/sync/src/class-functions.php]: I think that it would be a bit more concise to change this `switch` to either:

* An array map of constant => hosting provider, with a loop and a return if there's a match
* An `if/else`

mostly because it isn't a real syntactic `switch` because it doesn't check for a variable.

File: packages/sync/src/class-functions.php [hunk 4]
```diff
@@ -161,22 +161,90 @@ public static function get_post_type_features() {
...
+	/**
+	 * Return a hosting provider using a set of known classes.
+	 *
+	 * @return mixed A host identifier string or false.
+	 */
+	public function get_hosting_provider_by_known_class() {
+		$hosting_provider = false;
+
+		switch ( true ) {
+			case ( class_exists( '\\WPaaS\\Plugin' ) ):
```
Comment (tyxla) [inline packages/sync/src/class-functions.php]: The same comment from above applies for this `switch`, too.

File: packages/sync/src/class-functions.php [hunk 5]
```diff
@@ -161,22 +161,90 @@ public static function get_post_type_features() {
...
+
+	/**
+	 * Return a hosting provider using a set of known functions.
+	 *
+	 * @return mixed A host identifier string or false.
+	 */
+	public function get_hosting_provider_by_known_function() {
+		$hosting_provider = false;
+
+		switch ( true ) {
```
Comment (tyxla) [inline packages/sync/src/class-functions.php]: The same comment from above applies for this `switch`, too.

**Review by tyxla — DISMISSED** [review]
This looks good from my perspective 👍 

I'd only suggest against the way that `switch(true)` is used, but I don't have super strong feelings if y'all prefer it to be that way.

File: packages/sync/src/class-functions.php (hunk 3 above)
Comment (oikeme) [reply packages/sync/src/class-functions.php]: Good call. I've changed this. Updated previously due to not using the Constants class but this is definitely more concise with the loop.

File: packages/sync/src/class-functions.php [hunk 6]
```diff
@@ -161,22 +161,83 @@ public static function get_post_type_features() {
...
+			'PAGELYBIN'            => 'pagely',
+			'KINSTAMU_VERSION'     => 'kinsta',
+			'FLYWHEEL_CONFIG_DIR'  => 'flywheel',
+			'IS_PRESSABLE'         => 'pressable',
+			'VIP_GO_ENV'           => 'vip-go',
+		);
+
+		foreach ( $hosting_provider_constants as $constant => $constant_value ) {
+			if ( Constants::is_defined( $constant ) ) {
+				if ( 'VIP_GO_ENV' === $constant && false === Constants::get_constant( 'VIP_GO_ENV' ) ) {
```
Comment (ebinnion) [inline packages/sync/src/class-functions.php]: I wonder if it makes sense to apply this `false` check to all of the constants? Maybe something like:

```
if ( Constants::is_defined( $constant ) && false !== Constants::get_constant( 'VIP_GO_ENV' )  ) {
  return $constant_value;
}
```

This isn't a blocker, but it could minimize the need for us to change this in the future.

File: packages/sync/src/class-functions.php [hunk 7]
```diff
@@ -161,22 +161,83 @@ public static function get_post_type_features() {
...
+	/**
+	 * Return a hosting provider using a set of known functions.
+	 *
+	 * @return mixed A host identifier string or false.
+	 */
+	public function get_hosting_provider_by_known_function() {
+		$hosting_provider = false;
+
+		switch ( true ) {
+			case ( function_exists( 'is_wpe' ) || function_exists( 'is_wpe_snapshot' ) ):
```
Comment (ebinnion) [inline packages/sync/src/class-functions.php]: Personally, I'd prefer this as two separate case statements.


```
case function_exists( 'is_wpe' ):
case  function_exists( 'is_wpe_snapshot' ) ):
  $hosting_provider = 'wpe';
  break;
```

But, that's just a personal preference. We don't need to change it here.

**Review by ebinnion — APPROVED** [review]
I left a couple of comments, but they're not blockers. I like the change to use an array instead of a case for checking constants. It seems to have simplified the code quite a bit. 

I think it's fine to not make the same change to the other methods for now at least. Instead, we can make that change if/when necessary, as we add more functions and classes to check.

File: packages/sync/src/class-functions.php (hunk 6 above)
Comment (oikeme) [reply packages/sync/src/class-functions.php]: Wasn't sure on this as this logic is from before the update and specific to vip-go but it's a good call. I've updated this for everything.

File: packages/sync/src/class-functions.php (hunk 7 above)
Comment (oikeme) [reply packages/sync/src/class-functions.php]: I missed this in the prior commit but switched this to a loop instead

File: packages/sync/src/class-functions.php (hunk 6 above)
Comment (tyxla) [reply packages/sync/src/class-functions.php]: Seems like this still needs to be updated though as it's `VIP_GO_ENV`-specific?

**Review by ebinnion — APPROVED** [review]

**Comment by oikeme** [comment]
@tyxla I had a comment from you in an email earlier today about an issue I think may have been resolved. Can you confirm?

**Review by jeherve — APPROVED** [review]
This looks good to me! 👍

**Comment by tyxla** [comment]
I can see some of the comments were marked as resolved but not addressed, but they were also not blockers. So feel free to 🚢 if it's all intended 👍
